pkg/bucket/bandwidth/measurement.go

```diff
@@ -0,0 +1,133 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bandwidth
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// betaBucket is the weight used to calculate exponential moving average
+	betaBucket = 0.1 // Number of averages considered = 1/(1-betaObject)
+)
+
+// bucketMeasurement captures the bandwidth details for one bucket
+type bucketMeasurement struct {
+	lock                       sync.Mutex
+	cumulativeBytes            uint64        // Sum of all bytes transferred before the latest update
+	cumulativeTime             time.Duration // Total time taken for bytes
+	sumOfTimePerObject         time.Duration // Sum of time taken by objects in a window
+	sumOfTimePerObjectInWindow time.Duration // Sum of time taken by objects in a window
+	bytesSinceLastWindow       uint64        // Total bytes since last window was processed
+	windowStart                time.Time     // Start time for window
+	expMovingAvg               float64       // Previously calculate sliding window
+}
+
+// newBucketMeasurement creates a new instance of the measurement with the initial start time.
+func newBucketMeasurement(initTime time.Time) *bucketMeasurement {
+	return &bucketMeasurement{
+		windowStart: initTime,
+	}
+}
+
+// incrementBytes add bytes reported for a bucket.
+func (m *bucketMeasurement) incrementBytes(bytes uint64) {
+	atomic.AddUint64(&m.cumulativeBytes, bytes)
+	atomic.AddUint64(&m.bytesSinceLastWindow, bytes)
+}
+
+// calculateWindow processes the measurements between fixed time slots.
+func (m *bucketMeasurement) calculateWindow(windowEnd time.Time) {
+	// Calculate aggregate avg bandwidth and exp window avg
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	defer func() {
+		m.windowStart = windowEnd
+		m.sumOfTimePerObjectInWindow = 0
+	}()
+
+	if windowEnd.Before(m.windowStart) {
+		return
+	}
+
+	duration := windowEnd.Sub(m.windowStart)
+
+	// Do not update cumulative time if no traffic
+	bytesSinceLastWindow := atomic.SwapUint64(&m.bytesSinceLastWindow, 0)
+	if bytesSinceLastWindow > 0 {
+		m.cumulativeTime += duration
+		m.sumOfTimePerObject += m.sumOfTimePerObjectInWindow
+		// If total time spend copying object is less than wall clock time, then use
+		// sum of object to avoid  under reporting bandwidth.
+		if m.sumOfTimePerObject > 0 && m.cumulativeTime > m.sumOfTimePerObject {
+			m.cumulativeTime = m.sumOfTimePerObject
+		}
+	}
+	if m.sumOfTimePerObjectInWindow > 0 && duration > m.sumOfTimePerObjectInWindow {
+		duration = m.sumOfTimePerObjectInWindow
```

Comment:
AFAICT this can be 0. Will crash the divisions later.

Reply:
There is a check for `time.Before()` and a check for `sumOfTimePerObjectInWindow>0`